### PR TITLE
added ensemble core and ensemble tutorial

### PR DESCRIPTION
**Pull request by TgithubJ**
# Pull Request Checklist

Before sending your pull requests, make sure you followed this list.

- Read [contributing guidelines](CONTRIBUTING.md).
- Check if my changes are consistent with the [guidelines](TBD).
- Changes are consistent with the [Coding Style](TBD).
- Run [Unit Tests](TBD).

Checklist for Pull Request
- Describe what this PR purpose. For example, for new features describe what problem this PR solves. If there is JIRA task, please include JIRA link. 
- When there are more than one tasks in the PR, use to do list in github mark down. 
- Include unit tests. If small test files are needed for your unit tests, please include them as well. Be conservative in adding test files, try to make them small and local instead of putting it on S3 or external database.
- Assign reviewer.

# Running unit tests
Using tools and libraries installed directly on your system.
   For example, to run all tests under graphengine, do:

   ```bash
   nose2
   ```

### CODE REVIEW

File: h1st/h1st/core/ensemble.py [hunk 1]
```diff
@@ -2,106 +2,71 @@
...
 
 
 class StackEnsemble(Model):
-    def __init__(self, models, ensembler):
-        self.models = models
-        # must have .fit() / .predict(), currently assumes sklearn
-        self.model = ensembler
+    def __init__(self, ensembler, sub_models):
+        self.metrics = {}
+        self.ensembler = ensembler
```
Comment (thangtp) [inline h1st/h1st/core/ensemble.py:16]: @TgithubJ "ensemble" is the private variable, please use _ensemble. Also for "metrics" variable

File: h1st/h1st/core/ensemble.py [hunk 2]
```diff
@@ -2,106 +2,71 @@
...
 
 from h1st.core.model import Model
 
 
 class StackEnsemble(Model):
-    def __init__(self, models, ensembler):
-        self.models = models
-        # must have .fit() / .predict(), currently assumes sklearn
-        self.model = ensembler
+    def __init__(self, ensembler, sub_models):
```
Comment (thangtp) [inline h1st/h1st/core/ensemble.py:14]: pls consider adding typing for parameters. Ex: def __init__(self, ensembler: Model, sub_models: List[Model])

File: h1st/h1st/core/ensemble.py [hunk 3]
```diff
@@ -2,106 +2,71 @@
...
+        self.ensembler.fit(X_train, y_train)
+
+    def predict(self, data: dict):
+        X = data['X']
+        X = self.preprocess(X)
+        return {'predictions': self.ensembler.predict(X)}
+
+
+class StackEnsembleClassifier(StackEnsemble):
+    def evaluate(self, data: dict, metrics=['accuracy']):
```
Comment (thangtp) [inline h1st/h1st/core/ensemble.py:54]: to be consistent, please use List/Dict from typing module

File: h1st/h1st/core/ensemble.py [hunk 4]
```diff
@@ -2,106 +2,71 @@
...
-            else:
-                all_labels = np.vstack((all_labels, labels))
-
-        # assumption: all_inputs ~ numpy array, all_labels ~ numpy array
-        self.model.fit(all_inputs, all_labels)
-
-
-class MultiOutputClassifierEnsemble(StackEnsemble):
-    def __init__(self, models):
+        print(X.shape, predictions[0].shape, predictions[1].shape)
```
Comment (thangtp) [inline h1st/h1st/core/ensemble.py:26]: use logging instead

File: h1st/h1st/core/ensemble.py [hunk 5]
```diff
@@ -2,106 +2,71 @@
...
+
+
+class StackEnsembleClassifier(StackEnsemble):
+    def evaluate(self, data: dict, metrics=['accuracy']):
+        X_test = data['X_test']
+        y_test = data['y_test']
+        y_pred = self.predict({'X': X_test})['predictions']
+        # print(y_test.shape, y_pred.shape)
+        self.metrics['confusion_matrix'] = confusion_matrix(y_test, y_pred)
+        self.metrics['recall'], self.metrics['recall'], self.metrics['f1'] , self.metrics['support'] \
```
Comment (thangtp) [inline h1st/h1st/core/ensemble.py:60]: I am not sure whether this is what you want it to be or a typo issue.
self.metrics['recall'], self.metrics['recall']

If it is a type issue, it should be 
self.metrics['precision'], self.metrics['recall']

File: h1st/h1st/core/ensemble.py [hunk 6]
```diff
@@ -2,106 +2,71 @@
...
+    #     X_train = data['X_train']
+    #     y_train = data['y_train']
+    #     return {'X_train': X_train, 'y_train': y_train}
+
+    def train(self, prepared_data: dict):
+        X_train, y_train = (prepared_data['X_train'], prepared_data['y_train'])
+        X_train = self.preprocess(X_train, training=True)
+        self.ensembler.fit(X_train, y_train)
+
+    def predict(self, data: dict):
```
Comment (thangtp) [inline h1st/h1st/core/ensemble.py:47]: pls add return type to signature

**Comment by thangtp** [comment]
General comment: please also add docstrings for all public framework functions / modules @TgithubJ

**Comment by TgithubJ** [comment]
Old code structure. Will fork again and PR.

**Pull request closed without merging**